test/distributed/org/apache/cassandra/distributed/test/guardrails/GuardrailCollectionSizeOnSSTableWriteTest.java

```diff
@@ -0,0 +1,438 @@
+
+        if (driverCluster != null)
+            driverCluster.close();
+
+        if (cluster != null)
+            cluster.close();
+    }
+
+    @Override
+    protected Cluster getCluster()
+    {
+        return cluster;
+    }
+
+    @Test
+    public void testSetSize() throws Throwable
+    {
+        schemaChange("CREATE TABLE %s (k int PRIMARY KEY, v set<blob>)");
+
+        execute("INSERT INTO %s (k, v) VALUES (0, null)");
+        execute("INSERT INTO %s (k, v) VALUES (1, ?)", set());
+        execute("INSERT INTO %s (k, v) VALUES (2, ?)", set(allocate(1)));
+        execute("INSERT INTO %s (k, v) VALUES (3, ?)", set(allocate(WARN_THRESHOLD / 2)));
+        assertNotWarnedOnFlush();
+
+        execute("INSERT INTO %s (k, v) VALUES (4, ?)", set(allocate(WARN_THRESHOLD)));
+        assertWarnedOnFlush(warnMessage("4"));
+
+        execute("INSERT INTO %s (k, v) VALUES (5, ?)", set(allocate(WARN_THRESHOLD / 4), allocate(WARN_THRESHOLD * 3 / 4)));
+        assertWarnedOnFlush(warnMessage("5"));
+
+        execute("INSERT INTO %s (k, v) VALUES (6, ?)", set(allocate(FAIL_THRESHOLD)));
+        assertFailedOnFlush(failMessage("6"));
+
+        execute("INSERT INTO %s (k, v) VALUES (7, ?)", set(allocate(FAIL_THRESHOLD / 4), allocate(FAIL_THRESHOLD * 3 / 4)));
+        assertFailedOnFlush(failMessage("7"));
+    }
+
+    @Test
+    public void testSetSizeFrozen()
+    {
+        schemaChange("CREATE TABLE %s (k int PRIMARY KEY, v frozen<set<blob>>)");
+
+        execute("INSERT INTO %s (k, v) VALUES (0, null)");
+        execute("INSERT INTO %s (k, v) VALUES (1, ?)", set());
+        execute("INSERT INTO %s (k, v) VALUES (2, ?)", set(allocate(1)));
+        execute("INSERT INTO %s (k, v) VALUES (4, ?)", set(allocate(WARN_THRESHOLD)));
+        execute("INSERT INTO %s (k, v) VALUES (5, ?)", set(allocate(FAIL_THRESHOLD)));
+
+        // frozen collections size is not checked during sstable write
+        assertNotWarnedOnFlush();
+    }
+
+    @Test
+    public void testSetSizeWithUpdates()
+    {
+        schemaChange("CREATE TABLE %s (k int PRIMARY KEY, v set<blob>)");
+
+        execute("INSERT INTO %s (k, v) VALUES (0, ?)", set(allocate(1)));
+        execute("UPDATE %s SET v = v + ? WHERE k = 0", set(allocate(1)));
+        assertNotWarnedOnFlush();
+
+        execute("INSERT INTO %s (k, v) VALUES (1, ?)", set(allocate(WARN_THRESHOLD / 4)));
+        execute("UPDATE %s SET v = v + ? WHERE k = 1", set(allocate(WARN_THRESHOLD * 3 / 4)));
+        assertWarnedOnFlush(warnMessage("1"));
+
+        execute("INSERT INTO %s (k, v) VALUES (2, ?)", set(allocate(FAIL_THRESHOLD / 4)));
+        execute("UPDATE %s SET v = v + ? WHERE k = 2", set(allocate(FAIL_THRESHOLD * 3 / 4)));
+        assertFailedOnFlush(failMessage("2"));
+
+        execute("INSERT INTO %s (k, v) VALUES (4, ?)", set(allocate(FAIL_THRESHOLD)));
+        execute("UPDATE %s SET v = v - ? WHERE k = 4", set(allocate(FAIL_THRESHOLD)));
+        assertNotWarnedOnFlush();
```

Comment:
Do we want to add a check it failed here?

Reply:
Do you mean checking that the two `execute` calls fail because each of them is inserting a collection over the size threshold? Those queries aren't expected to fail in dtests because we are using a superuser, and superusers are excluded from guardrail checks. This dtest class is focused only on checks done during flushes and writes, whereas the checks done at query time are verified on `GuardrailCollectionSizeTest`.

Reply:
Shouldn't the insert in line [141](https://github.com/apache/cassandra/pull/1459/commits/6acad39969107be2b7ac44269f1c178ad37fae5d#diff-e928ebcf99e785567fcfa945c1af327c85ec0e7c50681042969357881886a411R141) fail the same way it failed [here](https://github.com/apache/cassandra/pull/1459/commits/6acad39969107be2b7ac44269f1c178ad37fae5d#diff-e928ebcf99e785567fcfa945c1af327c85ec0e7c50681042969357881886a411R102)?

If you insert `FAIL_THRESHOLD` I would expect it to fail at sstable level. I tried and the test seems to pass:
```
execute("INSERT INTO %s (k, v) VALUES (4, ?)", set(allocate(FAIL_THRESHOLD)));
        assertFailedOnFlush(failMessage("4"));
        execute("UPDATE %s SET v = v - ? WHERE k = 4", set(allocate(FAIL_THRESHOLD)));
        assertNotWarnedOnFlush();
```

Reply:
The `INSERT` on line 141 doesn't fail at query time because dtests use a superuser that is excluded from guardrails, although the guardrail would be triggered on the next flush, as it's shown by your example above.

The `UPDATE` on line 142 never triggers the guardrail at query time nor at flush time because it's a deletion inserting tombstones, and the guardrail doesn't count them as collection elements by design.

However, if we put the `INSERT` and the `UPDATE` together, so the writes go to the same memtable, then the deletions in the `UPDATE` will counteract the writes in the `INSERT`, so there won't be an activation of the guardrail when the memtable is flushed to a new sstable. 

Note that `allocate(FAIL_THRESHOLD)` always returns the same element, so the `UPDATE` is deleting what was written by the `INSERT`. Results would be different if we delete a different item, for example:
```
execute("INSERT INTO %s (k, v) VALUES (4, ?)", set(allocate(FAIL_THRESHOLD)));
execute("UPDATE %s SET v = v - ? WHERE k = 4", set(allocate(FAIL_THRESHOLD + 1))); // 1 extra byte
assertFailedOnFlush(failMessage("4"));
```
Or even:
```
execute("INSERT INTO %s (k, v) VALUES (4, ?)", set(allocate(FAIL_THRESHOLD)));
execute("UPDATE %s SET v = v - ? WHERE k = 4", set());
assertFailedOnFlush(failMessage("4"));
```

Reply:
Correct, it is the _flush_ on `assertFailedOnFlush(failMessage("4"));` that does the trick, agreed.